components/script/animations.rs

```diff
@@ -0,0 +1,290 @@
+        update: &mut AnimationsUpdate,
+    ) {
+        for animation in set.animations.iter_mut() {
+            if animation.state == AnimationState::Running && animation.has_ended(now) {
+                animation.state = AnimationState::Finished;
+                update.add_event(
+                    animation.node,
+                    animation.name.to_string(),
+                    TransitionOrAnimationEventType::AnimationEnd,
+                    animation.active_duration(),
+                );
+            }
+        }
+
+        for transition in set.transitions.iter_mut() {
+            if transition.state == AnimationState::Running && transition.has_ended(now) {
+                transition.state = AnimationState::Finished;
+                update.add_event(
+                    transition.node,
+                    transition.property_animation.property_name().into(),
+                    TransitionOrAnimationEventType::TransitionEnd,
+                    transition.property_animation.duration,
+                );
+            }
+        }
+    }
+
+    /// Send events for canceled animations. Currently this only handles canceled
+    /// transitions, but eventually this should handle canceled CSS animations as
+    /// well.
+    fn handle_canceled_animations(
+        set: &mut ElementAnimationSet,
+        now: f64,
+        update: &mut AnimationsUpdate,
+    ) {
+        for transition in &set.transitions {
+            if transition.state == AnimationState::Canceled {
+                // TODO(mrobinson): We need to properly compute the elapsed_time here
+                // according to https://drafts.csswg.org/css-transitions/#event-transitionevent
+                update.add_event(
+                    transition.node,
+                    transition.property_animation.property_name().into(),
+                    TransitionOrAnimationEventType::TransitionCancel,
+                    (now - transition.start_time).max(0.),
+                );
+            }
+        }
+
+        // TODO(mrobinson): We need to send animationcancel events.
+        set.clear_canceled_animations();
+    }
+
+    fn handle_new_animations(set: &mut ElementAnimationSet, update: &mut AnimationsUpdate) {
+        for animation in set.animations.iter_mut() {
+            animation.is_new = false;
+        }
+
+        for transition in set.transitions.iter_mut() {
+            if transition.is_new {
+                // TODO(mrobinson): We need to properly compute the elapsed_time here
+                // according to https://drafts.csswg.org/css-transitions/#event-transitionevent
+                update.add_event(
+                    transition.node,
+                    transition.property_animation.property_name().into(),
+                    TransitionOrAnimationEventType::TransitionRun,
+                    0.,
+                );
+                transition.is_new = false;
+            }
+        }
+    }
+}
+
+impl MallocSizeOf for Animations {
```

Comment:
I'm not sure this works because `MallocSizeOf` doesn't know what to do with `ServoArc`. I also cannot implement `MallocSizeOf` for `ServoArc` because it's defined outside the crate.

```
168 | impl MallocSizeOf for Arc<RwLock<FxHashMap<OpaqueNode, ElementAnimationSet>>> {
    | ^^^^^^^^^^^^^^^^^^^^^^-------------------------------------------------------
    | |                     |
    | |                     `style::servo_arc::Arc` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
```

I have modified this to include the other member of the `Animations` though. Let me know if there's a better way to accomplish this.

Reply:
Good point; Arcs always cause us problems. Let's add this comment above Animations:
`// Make sure to update the MallocSizeOf implementation when changing the contents of this struct.`